File: diagnostics_tools/src/periodic_event_status.cpp
/*********************************************************************
 * Software License Agreement (BSD License)
 *
 *  Copyright (c) 2020, QinetiQ, Inc.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of QinetiQ nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/

#include "diagnostic_tools/periodic_event_status.h"

#include <cinttypes>
#include <string>

#include <diagnostic_msgs/DiagnosticStatus.h>

namespace qna
{
namespace diagnostic_tools
{
PeriodicEventStatus::PeriodicEventStatus(const std::string& name)
  : PeriodicEventStatus(name, PeriodicEventStatusParams{}) {}

PeriodicEventStatus::PeriodicEventStatus(const std::string& name, PeriodicEventStatusParams params)
  : PeriodicDiagnosticTask(name),
    params_(std::move(params)),
    short_term_period_(params.short_term_avg_window()),
    long_term_period_(params.long_term_avg_window()),
    start_stamp_(ros::Time::now()) {}

void PeriodicEventStatus::tick(const ros::Time& stamp)
{
  std::lock_guard<std::mutex> guard(mutex_);
  if (!last_stamp_.isZero())
  {
    if (last_stamp_ <= stamp)
    {
      const double time_delta = (stamp - last_stamp_).toSec();
      if (time_delta <= params_.max_reasonable_period())
      {
        short_term_period_.update(time_delta);
        long_term_period_.update(time_delta);
      }
      else
      {
        ROS_DEBUG_NAMED(
          "diagnostics_tools",
          "Time delta %f too long, ignoring",
          time_delta);
      }
    }
    else
    {
      ROS_DEBUG_NAMED(
        "diagnostics_tools",
        "Time went backwards from %f to %f, ignoring",
        last_stamp_.toSec(), stamp.toSec());
    }
  }
  last_stamp_ = stamp;
}

void PeriodicEventStatus::run(diagnostic_updater::DiagnosticStatusWrapper& stat)
{
  std::lock_guard<std::mutex> guard(mutex_);
  const ros::Time current_stamp = ros::Time::now();
  const double time_delta = (current_stamp - last_stamp_).toSec();
  if (short_term_period_.sample_count() > 0 && time_delta < params_.max_reasonable_period())
  {
    if (short_term_period_.average() < params_.min_acceptable_period() ||
        short_term_period_.average() > params_.max_acceptable_period())
    {
      stat.summary(params_.abnormal_diagnostic().status(),
                   params_.abnormal_diagnostic().description());
      if (params_.abnormal_diagnostic().has_code())
      {
        stat.addf("Code", "%" PRIu64, params_.abnormal_diagnostic().code());
      }
    }
    else
    {
      stat.summary(params_.normal_diagnostic().status(),
                   params_.normal_diagnostic().description());
      if (params_.normal_diagnostic().has_code())
      {
        stat.addf("Code", "%" PRIu64, params_.normal_diagnostic().code());
      }
    }
    stat.addf("Average period (short term)", "%f", short_term_period_.average());
    stat.addf("Minimum period (short term)", "%f", short_term_period_.minimum());
    stat.addf("Maximum period (short term)", "%f", short_term_period_.maximum());
  }
  else
  {
    const double time_since_init = (current_stamp - start_stamp_).toSec();

    if (time_since_init <= params_.max_reasonable_period())
    {
      // Likely still initializing, skip diagnostic checks
      return;
    }

    // Event is likely stale, reset estimate
    short_term_period_.reset();

    stat.summary(params_.stale_diagnostic().status(),
                 params_.stale_diagnostic().description());
    if (params_.stale_diagnostic().has_code())
    {
      stat.addf("Code", "%" PRIu64, params_.stale_diagnostic().code());
    }
  }
  if (long_term_period_.sample_count() > 0)
  {
    stat.addf("Average period (long term)", "%f", long_term_period_.average());
    stat.addf("Minimum period (long term)", "%f", long_term_period_.minimum());
    stat.addf("Maximum period (long term)", "%f", long_term_period_.maximum());
  }
  stat.addf("Minimum acceptable period", "%f", params_.min_acceptable_period());
  stat.addf("Maximum acceptable period", "%f", params_.max_acceptable_period());
}

}  // namespace diagnostic_tools
}  // namespace qna

File: diagnostics_tools/src/diagnostic_tools/periodic_event_status.py
"""
 * Software License Agreement (BSD License)
 *
 *  Copyright (c) 2020, QinetiQ, Inc.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of QinetiQ nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
"""

import collections

import rospy

from diagnostic_tools.diagnostic import Diagnostic
from diagnostic_tools.periodic_diagnostic_task import PeriodicDiagnosticTask
from diagnostic_tools.sampled_statistics import SampledStatistics


class PeriodicEventStatus(PeriodicDiagnosticTask):

    Diagnostics = collections.namedtuple('Diagnostics', 'normal abnormal stale')
    Configuration = collections.namedtuple(
        'Configuration',
        ' '.join([
            'min_acceptable_period max_acceptable_period max_reasonable_period',
            'short_term_avg_window long_term_avg_window diagnostics',
        ])
    )

    @staticmethod
    def configure(
        min_acceptable_period=-1,
        max_acceptable_period=5,
        max_reasonable_period=None,
        short_term_avg_window=1,
        long_term_avg_window=10000,
        normal_diagnostic=Diagnostic.OK,
        abnormal_diagnostic=Diagnostic.WARN,
        stale_diagnostic=Diagnostic.STALE
    ):
        normal_diagnostic = Diagnostic.build(normal_diagnostic)
        if not normal_diagnostic.description:
            normal_diagnostic.description = 'Rate within tolerance'
        abnormal_diagnostic = Diagnostic.build(abnormal_diagnostic)
        if not abnormal_diagnostic.description:
            abnormal_diagnostic.description = 'Rate too high or too low'
        stale_diagnostic = Diagnostic.build(stale_diagnostic)
        if not stale_diagnostic.description:
            stale_diagnostic.description = 'Not enough data since last update'
        if max_reasonable_period is None:
            max_reasonable_period = 5 * max_acceptable_period
        return PeriodicEventStatus.Configuration(
            min_acceptable_period=min_acceptable_period,
            max_acceptable_period=max_acceptable_period,
            max_reasonable_period=max_reasonable_period,
            short_term_avg_window=short_term_avg_window,
            long_term_avg_window=long_term_avg_window,
            diagnostics=PeriodicEventStatus.Diagnostics(
                normal=normal_diagnostic,
                abnormal=abnormal_diagnostic,
                stale=stale_diagnostic
            )
        )

    def __init__(self, name='rate check', config=None, **kwargs):
        PeriodicDiagnosticTask.__init__(self, name, **kwargs)
        if config is None:
            config = PeriodicEventStatus.configure()
        self._config = config

        self._last_time = rospy.Time(0)
        self._short_term_period = SampledStatistics(
            self._config.short_term_avg_window, dtype=float
        )
        self._long_term_period = SampledStatistics(
            self._config.long_term_avg_window, dtype=float
        )

    def tick(self, time=None):
        time = time or rospy.Time.now()
        with self._lock:
            if not self._last_time.is_zero():
                if self._last_time <= time:
                    delta_in_seconds = (time - self._last_time).to_sec()
                    if delta_in_seconds <= self._config.max_reasonable_period:
                        self._short_term_period.update(delta_in_seconds)
                        self._long_term_period.update(delta_in_seconds)
                    else:
                        rospy.logdebug(
                            'Time delta %f too long, ignoring', delta_in_seconds
                        )
                else:
                    rospy.logdebug(
                        'Time went backwards from %f to %f, ignoring',
                        self._last_time.to_sec(), time.to_sec()
                    )
            self._last_time = time

    def run(self, stat):
        with self._lock:
            diagnostic = self._config.diagnostics.normal
            if rospy.Time.now() - self._last_time > self._config.max_reasonable_period:
                # Unreasonable time delta, event is likely stale.
                self._short_term_period.reset()
            if self._short_term_period.sample_count > 0:
                if self._last_cycle_period.average < self._config.min_acceptable_period or \
                   self._last_cycle_period.average > self._config.max_acceptable_period:
                    diagnostic = self._config.diagnostics.abnormal
                stat.summary(diagnostic.status, diagnostic.description)
                if diagnostic.code is not None:
                    stat.add('Code', diagnostic.code)
                stat.add(
                    'Average period (last cycle)', self._last_cycle_period.average)
                stat.add(
                    'Minimum period (last cycle)', self._last_cycle_period.minimum)
                stat.add(
                    'Maximum period (last cycle)', self._last_cycle_period.maximum)
            else:
                diagnostic = self._config.diagnostics.stale
                stat.summary(diagnostic.status, diagnostic.description)
                if diagnostic.code is not None:
                    stat.add('Code', diagnostic.code)
            if self._long_term_period.sample_count > 0:
                stat.add(
                    'Average period (historic)', self._long_term_period.average)
                stat.add(
                    'Minimum period (historic)', self._long_term_period.minimum)
                stat.add(
                    'Maximum period (historic)', self._long_term_period.maximum)
            stat.add('Minimum acceptable period', self._config.min_acceptable_period)
            stat.add('Maximum acceptable period', self._config.max_acceptable_period)
        return stat
